File: komodo/yamlite.py
"""The YAML subset task blocks use: flat keys, scalars, inline lists, dashed lists."""

from __future__ import annotations

import re
from typing import Any, Dict, List

KEY_LINE = re.compile(r"^([A-Za-z_][A-Za-z0-9_]*)\s*:\s*(.*)$")
DASH_LINE = re.compile(r"^\s*-\s+(.*)$")


class YamliteError(ValueError):
    """Raised when a block falls outside the supported subset."""


def _strip_comment(text: str) -> str:
    """Drops a trailing # comment unless the # sits inside quotes."""
    quote = None
    for index, char in enumerate(text):
        if quote:
            if char == quote:
                quote = None
            continue
        if char in "'\"":
            quote = char
            continue
        if char == "#" and (index == 0 or text[index - 1].isspace()):
            return text[:index].rstrip()
    return text.rstrip()


def _scalar(raw: str) -> Any:
    """Turns one token into str, int, bool, or None, honoring quotes."""
    text = raw.strip()
    if not text:
        return ""
    if len(text) >= 2 and text[0] == text[-1] and text[0] in "'\"":
        return text[1:-1]
    lowered = text.lower()
    if lowered in ("true", "yes"):
        return True
    if lowered in ("false", "no"):
        return False
    if lowered in ("null", "~"):
        return None
    if re.fullmatch(r"-?\d+", text):
        return int(text)
    return text


def _split_inline_list(body: str) -> List[str]:
    """Splits [a, b, "c, d"] on commas that sit outside quotes."""
    items, current, quote = [], [], None
    for char in body:
        if quote:
            current.append(char)
            if char == quote:
                quote = None
            continue
        if char in "'\"":
            quote = char
            current.append(char)
            continue
        if char == ",":
            items.append("".join(current))
            current = []
            continue
        current.append(char)
    tail = "".join(current)
    if tail.strip() or items:
        items.append(tail)
    return [item for item in items if item.strip()]


def loads(text: str) -> Dict[str, Any]:
    """Parses a block into a dict, raising YamliteError on anything outside the subset."""
    result: Dict[str, Any] = {}
    lines = text.splitlines()
    index = 0
    while index < len(lines):
        line = _strip_comment(lines[index])
        index += 1
        if not line.strip():
            continue
        if line[0].isspace():
            raise YamliteError("unexpected indentation at: %r" % line)
        match = KEY_LINE.match(line)
        if not match:
            raise YamliteError("expected 'key: value' at: %r" % line)
        key, value = match.group(1), match.group(2).strip()
        if value.startswith("[") and value.endswith("]"):
            result[key] = [_scalar(item) for item in _split_inline_list(value[1:-1])]
            continue
        if value:
            result[key] = _scalar(value)
            continue
        items: List[Any] = []
        while index < len(lines):
            candidate = _strip_comment(lines[index])
            if not candidate.strip():
                index += 1
                continue
            dash = DASH_LINE.match(candidate)
            if not dash:
                break
            items.append(_scalar(dash.group(1)))
            index += 1
        result[key] = items
    return result


def _quote(value: Any) -> str:
    """Renders a scalar, quoting strings the parser would otherwise misread."""
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        return str(value)
    text = str(value)
    needs_quote = (
        not text
        or text != text.strip()
        or any(char in text for char in ":#[]{},\"'")
        or text.lower() in ("true", "false", "yes", "no", "null", "~")
        or re.fullmatch(r"-?\d+", text) is not None
    )
    if needs_quote:
        return '"' + text.replace('"', '\\"') + '"'
    return text


def dumps(data: Dict[str, Any]) -> str:
    """Renders a dict back into the subset, lists dashed one per line."""
    lines = []
    for key, value in data.items():
        if isinstance(value, (list, tuple)):
            if not value:
                lines.append("%s: []" % key)
                continue
            lines.append("%s:" % key)
            lines.extend("  - %s" % _quote(item) for item in value)
            continue
        lines.append("%s: %s" % (key, _quote(value)))
    return "\n".join(lines) + ("\n" if lines else "")

File: internal/backlog/yamlite.go
package backlog

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// The YAML subset a task block uses: flat keys, scalars, inline lists, dashed lists.

var (
	keyLine  = regexp.MustCompile(`^([A-Za-z_][A-Za-z0-9_]*)\s*:\s*(.*)$`)
	dashLine = regexp.MustCompile(`^\s*-\s+(.*)$`)
	intOnly  = regexp.MustCompile(`^-?\d+$`)
)

// Fields is one parsed block: keys in file order, values as string, int, bool, nil, or a slice.
type Fields struct {
	keys   []string
	values map[string]any
}

// Get returns the value stored under key, and whether the key was present.
func (f Fields) Get(key string) (any, bool) {
	value, ok := f.values[key]
	return value, ok
}

// Keys returns the block's keys in the order they were written.
func (f Fields) Keys() []string { return append([]string(nil), f.keys...) }

// Set stores a value, appending the key when it is new.
func (f *Fields) Set(key string, value any) {
	if f.values == nil {
		f.values = map[string]any{}
	}
	if _, seen := f.values[key]; !seen {
		f.keys = append(f.keys, key)
	}
	f.values[key] = value
}

// String returns the value under key as text, or the empty string.
func (f Fields) String(key string) string {
	value, ok := f.values[key]
	if !ok || value == nil {
		return ""
	}
	if text, ok := value.(string); ok {
		return text
	}
	return fmt.Sprint(value)
}

// List returns the value under key as a slice of strings, or nil.
func (f Fields) List(key string) []string {
	value, ok := f.values[key]
	if !ok {
		return nil
	}
	items, ok := value.([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(items))
	for _, item := range items {
		if item == nil {
			continue
		}
		out = append(out, fmt.Sprint(item))
	}
	return out
}

// stripComment drops a trailing # comment unless the # sits inside quotes.
func stripComment(text string) string {
	var quote rune
	for index, char := range text {
		if quote != 0 {
			if char == quote {
				quote = 0
			}
			continue
		}
		if char == '\'' || char == '"' {
			quote = char
			continue
		}
		if char == '#' && (index == 0 || isSpaceByte(text[index-1])) {
			return strings.TrimRight(text[:index], " \t")
		}
	}
	return strings.TrimRight(text, " \t")
}

// isSpaceByte reports whether the byte is an ASCII space or tab.
func isSpaceByte(b byte) bool { return b == ' ' || b == '\t' }

// scalar turns one token into a string, int, bool, or nil, honouring quotes.
func scalar(raw string) any {
	text := strings.TrimSpace(raw)
	if text == "" {
		return ""
	}
	if len(text) >= 2 && text[0] == text[len(text)-1] && text[0] == '\'' {
		// Inside single quotes YAML writes a single quote twice.
		return strings.ReplaceAll(text[1:len(text)-1], "''", "'")
	}
	if len(text) >= 2 && text[0] == text[len(text)-1] && text[0] == '"' {
		return text[1 : len(text)-1]
	}
	switch strings.ToLower(text) {
	case "true", "yes":
		return true
	case "false", "no":
		return false
	case "null", "~":
		return nil
	}
	if intOnly.MatchString(text) {
		number, err := strconv.Atoi(text)
		if err == nil {
			return number
		}
	}
	return text
}

// splitInline splits [a, b, "c, d"] on the commas that sit outside quotes.
func splitInline(body string) []string {
	var items []string
	var current strings.Builder
	var quote rune
	for _, char := range body {
		if quote != 0 {
			current.WriteRune(char)
			if char == quote {
				quote = 0
			}
			continue
		}
		if char == '\'' || char == '"' {
			quote = char
			current.WriteRune(char)
			continue
		}
		if char == ',' {
			items = append(items, current.String())
			current.Reset()
			continue
		}
		current.WriteRune(char)
	}
	if tail := current.String(); strings.TrimSpace(tail) != "" || len(items) > 0 {
		items = append(items, tail)
	}
	out := items[:0]
	for _, item := range items {
		if strings.TrimSpace(item) != "" {
			out = append(out, item)
		}
	}
	return out
}

// ParseFields reads a block into Fields, failing on anything outside the subset.
func ParseFields(text string) (Fields, error) {
	fields := Fields{values: map[string]any{}}
	lines := strings.Split(text, "\n")
	index := 0
	for index < len(lines) {
		line := stripComment(lines[index])
		index++
		if strings.TrimSpace(line) == "" {
			continue
		}
		if isSpaceByte(line[0]) {
			return fields, fmt.Errorf("unexpected indentation at: %q", line)
		}
		match := keyLine.FindStringSubmatch(line)
		if match == nil {
			return fields, fmt.Errorf("expected 'key: value' at: %q", line)
		}
		key, value := match[1], strings.TrimSpace(match[2])
		if strings.HasPrefix(value, "[") && strings.HasSuffix(value, "]") {
			var items []any
			for _, item := range splitInline(value[1 : len(value)-1]) {
				items = append(items, scalar(item))
			}
			if items == nil {
				items = []any{}
			}
			fields.Set(key, items)
			continue
		}
		if value != "" {
			fields.Set(key, scalar(value))
			continue
		}
		items := []any{}
		for index < len(lines) {
			candidate := stripComment(lines[index])
			if strings.TrimSpace(candidate) == "" {
				index++
				continue
			}
			dash := dashLine.FindStringSubmatch(candidate)
			if dash == nil {
				break
			}
			items = append(items, scalar(dash[1]))
			index++
		}
		fields.Set(key, items)
	}
	return fields, nil
}

// newlineRe matches any line break so quote can collapse one into a space.
var newlineRe = regexp.MustCompile(`\r\n|\r|\n`)

// quote renders a scalar, quoting text the parser would otherwise misread.
// A newline is collapsed to a space, since a raw one would break the fenced block.
func quote(value any) string {
	switch typed := value.(type) {
	case nil:
		return "null"
	case bool:
		if typed {
			return "true"
		}
		return "false"
	case int:
		return strconv.Itoa(typed)
	}
	text := newlineRe.ReplaceAllString(fmt.Sprint(value), " ")
	needs := text == "" || text != strings.TrimSpace(text) ||
		strings.ContainsAny(text, ":#[]{},\"'") || intOnly.MatchString(text)
	switch strings.ToLower(text) {
	case "true", "false", "yes", "no", "null", "~":
		needs = true
	}
	if !needs {
		return text
	}
	// A value holding a double quote takes single quotes, each single quote inside written twice.
	if strings.Contains(text, "\"") {
		return "'" + strings.ReplaceAll(text, "'", "''") + "'"
	}
	return `"` + text + `"`
}

// DumpFields renders Fields back into the subset, lists dashed one item per line.
func DumpFields(fields Fields) string {
	var out []string
	for _, key := range fields.keys {
		value := fields.values[key]
		items, isList := value.([]any)
		if !isList {
			out = append(out, key+": "+quote(value))
			continue
		}
		if len(items) == 0 {
			out = append(out, key+": []")
			continue
		}
		out = append(out, key+":")
		for _, item := range items {
			out = append(out, "  - "+quote(item))
		}
	}
	if len(out) == 0 {
		return ""
	}
	return strings.Join(out, "\n") + "\n"
}
